mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java

```diff
@@ -101,7 +111,20 @@
 
     @Override
     public Publisher<Void> delete(User user, MessageId messageId) {
-        throw new NotImplementedException("Will be implemented later");
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        return Flux.from(messageMetadataVault.listRelatedBuckets())
+            .flatMap(bucketName -> deleteFromBucket(bucketName, user, messageId))
+            .subscribeOn(Schedulers.elastic());
+    }
+
+    private Mono<Void> deleteFromBucket(BucketName bucketName, User user, MessageId messageId) {
+        return Mono.from(messageMetadataVault.retrieveStorageInformation(user, messageId))
+            .flatMap(storageInformation -> Mono.from(messageMetadataVault.remove(bucketName, user, messageId))
+                .thenReturn(storageInformation))
+            .flatMap(storageInformation -> blobStore.delete(bucketName, storageInformation.getBlobId()))
+            .then();
```

Comment:
I don't think you understood Benoit's point. I think you should call this directly in your delete method. You can get the bucket name from the storage info, thus you don't need to check in all buckets if the message is there or not (one of his points).  I think Benoit meant something more like this :
```
public Publisher<Void> delete(User user, MessageId messageId) {
 Preconditions.checkNotNull(user);
 Preconditions.checkNotNull(messageId);

 return Mono.from(metadatavault.retrieveStorageInformation(user, messageId)
   .flatMap(storageInformation -> metadatavault.remove(storageInformation.getBucket(), user, messageId)
       .then(Mono.just(storageInformation)))
   .flatMap(storageInformation -> blobStore.delete(storageInformation.getBucket(), storageInformation.getMessageId())));
}

Reply:
sorry, miss reading